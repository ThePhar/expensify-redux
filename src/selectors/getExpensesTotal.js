// (expenses) => adds up each expense in array and returns sum.

export default (expenses) => (
    expenses.reduce((accumulator, expense) => (
        accumulator + expense.amount
    ), 0)
);