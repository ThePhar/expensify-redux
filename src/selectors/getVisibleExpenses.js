import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return b.createdAt - a.createdAt;
        } else {
            return b.amount - a.amount;
        }
    });
};
