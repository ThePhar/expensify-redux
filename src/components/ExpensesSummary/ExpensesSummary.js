import numeral            from 'numeral';
import React              from 'react';
import { connect }        from 'react-redux';
import getExpensesTotal   from '../../selectors/getExpensesTotal';
import getVisibleExpenses from '../../selectors/getVisibleExpenses';

export const ExpensesSummary = (props) => {
    const pluralized = props.expenseCount === 1 ? 'expense' : 'expenses';
    const numberOfExpenses = props.expenseCount;
    const totalAmount = numeral(props.expenseTotal / 100).format('$0,0.00');

    return (
        <div className="ma3 br3 flex items-center justify-center pa4 bg-lightest-blue navy">
            <svg className="w1" data-icon="info" viewBox="0 0 32 32" style={{ fill: 'currentcolor' }}>
                <title>info icon</title>
                <path
                    d="M16 0 A16 16 0 0 1 16 32 A16 16 0 0 1 16 0 M19 15 L13 15 L13 26 L19 26 z M16 6 A3 3 0 0 0 16 12 A3 3 0 0 0 16 6" />
            </svg>
            <span className="lh-title ml3">Viewing {numberOfExpenses} {pluralized} totalling {totalAmount}</span>
        </div>
    );
};


// {/*<p>*/ }
// {/*    Viewing {props.expenseCount}*/}
// {/*    {*/}
// {/*        props.expenseCount === 1 ? ' expense ' : ' expenses '*/}
// {/*    }*/}
// {/*    totalling {numeral(props.expenseTotal / 100).format('$0,0.00')}*/}
// {/*</p>*/}

const mapStateToProps = (state) => ({
    expenseCount: getVisibleExpenses(state.expenses, state.filters).length,
    expenseTotal: getExpensesTotal(getVisibleExpenses(state.expenses, state.filters))
});

export default connect(mapStateToProps)(ExpensesSummary);