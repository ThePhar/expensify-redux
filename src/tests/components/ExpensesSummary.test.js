import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary/ExpensesSummary";
import expenses from "../fixtures/expenses";
import getVisibleExpenses from "../../selectors/getVisibleExpenses";
import getExpensesTotal from "../../selectors/getExpensesTotal";
import { filters } from "../fixtures/filters";

let wrapper, expenseCount, expenseTotal;
beforeEach(() => {
    expenseCount = getVisibleExpenses(expenses, filters).length;
    expenseTotal = getExpensesTotal(getVisibleExpenses(expenses, filters));
    wrapper = shallow(
        <ExpensesSummary
            expenseCount={expenseCount}
            expenseTotal={expenseTotal}
        />
    );
});

test("Should render summary of a single expense with singular noun.", () => {
    const singleExpenseList = [expenses[1]];
    wrapper.setProps({
        expenseCount: getVisibleExpenses(singleExpenseList, filters).length,
        expenseTotal: getExpensesTotal(getVisibleExpenses(singleExpenseList, filters))
    });
    expect(wrapper.html()).toContain("Viewing 1 expense totalling $1,095.00");
    expect(wrapper).toMatchSnapshot();
});

test("Should render summary of multiple expenses with plural noun.", () => {
    expect(wrapper.html()).toContain("Viewing 3 expenses totalling $1,141.95");
    expect(wrapper).toMatchSnapshot();
});