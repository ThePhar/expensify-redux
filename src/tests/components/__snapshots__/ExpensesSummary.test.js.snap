// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render summary of a single expense with singular noun. 1`] = `
<p>
  Viewing 
  1
   expense 
  totalling 
  $1,095.00
</p>
`;

exports[`Should render summary of multiple expenses with plural noun. 1`] = `
<p>
  Viewing 
  3
   expenses 
  totalling 
  $1,141.95
</p>
`;
